{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import concurrent.futures \n",
    "from multiprocessing.dummy import Pool as pool\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3825936/1364745899.py:1: UserWarning: loadtxt: Empty input file: \"/proj/htzhu/UKB_GWAS/results/ukb_phase1and2_roi/ukb_roi_volume_may12_2019_phase1and2_pheno99_allchr_withA2.txt\"\n",
      "  data = np.loadtxt('/proj/htzhu/UKB_GWAS/results/ukb_phase1and2_roi/ukb_roi_volume_may12_2019_phase1and2_pheno99_allchr_withA2.txt', dtype=str, delimiter=',')\n"
     ]
    }
   ],
   "source": [
    "data = np.loadtxt('/proj/htzhu/UKB_GWAS/results/ukb_phase1and2_roi/ukb_roi_volume_may12_2019_phase1and2_pheno99_allchr_withA2.txt', dtype=str, delimiter=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/proj/htzhu/UKB_GWAS/results/ukb_phase1and2_roi/ukb_roi_volume_may12_2019_phase1and2_pheno98_allchr_withA2.txt', 'r') as file:\n",
    "    lines = file.readlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CHR SNP BP A1 TEST  NMISS BETA STAT P \\tA2\\n'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lines[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['3', 'rs342909', '146341963', 'T', 'ADD', '18306', '-0.0699', '-6.594', '4.413e-11', 'C']\n",
      "['3', 'rs11399396', '146357520', 'T', 'ADD', '18316', '-0.0716', '-6.778', '1.256e-11', 'TA']\n",
      "['3', 'rs1674593', '146362450', 'A', 'ADD', '18514', '0.0643', '6.125', '9.269e-10', 'G']\n",
      "['3', 'rs1674592', '146362509', 'A', 'ADD', '18515', '0.06426', '6.121', '9.461e-10', 'T']\n",
      "['3', 'rs1702384', '146362636', 'C', 'ADD', '18515', '0.06426', '6.121', '9.461e-10', 'T']\n",
      "['3', 'rs36008198', '146362717', 'TGTA', 'ADD', '18458', '0.06499', '6.181', '6.515e-10', 'T']\n",
      "['3', 'rs1705839', '146362792', 'T', 'ADD', '18519', '0.06412', '6.109', '1.026e-09', 'G']\n",
      "['3', 'rs1705840', '146362864', 'T', 'ADD', '18519', '0.06412', '6.109', '1.026e-09', 'G']\n",
      "['3', 'rs1705841', '146362914', 'T', 'ADD', '18519', '0.06412', '6.109', '1.026e-09', 'G']\n",
      "['3', 'rs34803973', '146363009', 'A', 'ADD', '18105', '0.06473', '6.086', '1.179e-09', 'C']\n",
      "['3', 'rs386666824', '146363012', 'AG', 'ADD', '18076', '0.06513', '6.114', '9.927e-10', 'A']\n",
      "['3', 'rs1702385', '146363231', 'G', 'ADD', '18519', '0.06412', '6.109', '1.026e-09', 'T']\n",
      "['3', 'rs1702386', '146363400', 'C', 'ADD', '18522', '0.06383', '6.082', '1.208e-09', 'G']\n",
      "['3', 'rs1705842', '146363643', 'C', 'ADD', '18520', '0.0639', '6.089', '1.161e-09', 'T']\n",
      "['3', 'rs1705843', '146363667', 'C', 'ADD', '18522', '0.06383', '6.082', '1.208e-09', 'T']\n",
      "['3', 'rs9810961', '146363940', 'G', 'ADD', '18524', '0.06391', '6.09', '1.15e-09', 'A']\n",
      "['3', 'rs6799282', '146364127', 'G', 'ADD', '18524', '0.06395', '6.094', '1.123e-09', 'A']\n",
      "['3', 'rs6799360', '146364185', 'G', 'ADD', '18524', '0.06384', '6.083', '1.202e-09', 'A']\n",
      "['3', 'rs4541374', '146364201', 'G', 'ADD', '18524', '0.06373', '6.073', '1.279e-09', 'A']\n",
      "['3', 'rs1702391', '146364386', 'G', 'ADD', '18524', '0.06335', '6.037', '1.603e-09', 'A']\n",
      "['3', 'rs1674589', '146364424', 'A', 'ADD', '18524', '0.06314', '6.016', '1.823e-09', 'T']\n",
      "['3', 'rs1674591', '146364656', 'G', 'ADD', '18524', '0.06314', '6.016', '1.823e-09', 'A']\n",
      "['3', 'rs1627176', '146365115', 'T', 'ADD', '18524', '0.06335', '6.037', '1.603e-09', 'C']\n",
      "['3', 'rs1705844', '146365435', 'C', 'ADD', '18524', '0.06372', '6.072', '1.288e-09', 'A']\n",
      "['3', 'rs8179985', '146365527', 'T', 'ADD', '18525', '0.06374', '6.074', '1.271e-09', 'G']\n",
      "['3', 'rs1380168', '146366217', 'G', 'ADD', '18528', '0.06362', '6.064', '1.357e-09', 'T']\n",
      "['3', 'rs1380167', '146366361', 'T', 'ADD', '18527', '0.0636', '6.061', '1.376e-09', 'A']\n",
      "['3', 'rs1702390', '146366600', 'C', 'ADD', '18529', '0.06366', '6.067', '1.327e-09', 'T']\n",
      "['3', 'rs1702389', '146366635', 'T', 'ADD', '18529', '0.06366', '6.067', '1.327e-09', 'C']\n",
      "['3', 'rs112461436', '146366739', 'A', 'ADD', '18359', '-0.07145', '-6.823', '9.226e-12', 'AAAGTT']\n",
      "['3', 'rs1674587', '146367264', 'T', 'ADD', '18529', '0.06366', '6.067', '1.327e-09', 'A']\n",
      "['3', 'rs1674586', '146367278', 'G', 'ADD', '18529', '0.06366', '6.067', '1.327e-09', 'A']\n",
      "['3', 'rs1674585', '146367432', 'T', 'ADD', '18529', '0.06366', '6.067', '1.327e-09', 'A']\n",
      "['3', 'rs1702387', '146367603', 'G', 'ADD', '18529', '0.06366', '6.067', '1.327e-09', 'A']\n",
      "['3', 'rs1674584', '146367653', 'T', 'ADD', '18537', '0.06355', '6.057', '1.411e-09', 'C']\n",
      "['3', 'rs1613380', '146367804', 'A', 'ADD', '18566', '0.06355', '6.063', '1.359e-09', 'C']\n",
      "['3', 'rs9842222', '146368187', 'T', 'ADD', '18491', '0.06912', '5.86', '4.706e-09', 'C']\n",
      "['3', 'rs9847565', '146369166', 'G', 'ADD', '18272', '0.06518', '6.331', '2.499e-10', 'C']\n",
      "['3', 'rs12695824', '146369807', 'A', 'ADD', '18329', '-0.07424', '-7.065', '1.669e-12', 'T']\n",
      "['3', 'rs57946797', '146370310', 'A', 'ADD', '18240', '0.06828', '6.634', '3.351e-11', 'T']\n",
      "['3', 'rs4681329', '146371033', 'A', 'ADD', '18313', '-0.07434', '-7.068', '1.623e-12', 'G']\n",
      "['3', 'rs6762490', '146376451', 'A', 'ADD', '18284', '-0.07339', '-6.978', '3.093e-12', 'T']\n",
      "['3', 'rs4681330', '146378341', 'C', 'ADD', '18283', '-0.07377', '-7.014', '2.396e-12', 'T']\n",
      "['3', 'rs6796791', '146380850', 'T', 'ADD', '18283', '-0.07377', '-7.014', '2.396e-12', 'C']\n",
      "['3', 'rs9851764', '146382839', 'C', 'ADD', '18282', '-0.07374', '-7.012', '2.436e-12', 'A']\n",
      "['3', 'rs12633109', '146392141', 'T', 'ADD', '18244', '-0.0737', '-7.007', '2.518e-12', 'G']\n",
      "['13', 'rs303936', '72373889', 'C', 'ADD', '18472', '-0.06048', '-5.789', '7.183e-09', 'A']\n",
      "['13', 'rs5804584', '72375010', 'TA', 'ADD', '18433', '-0.06058', '-5.791', '7.117e-09', 'T']\n",
      "['13', 'rs5804585', '72375212', 'GT', 'ADD', '18431', '-0.06024', '-5.759', '8.578e-09', 'G']\n",
      "['13', 'rs303933', '72378343', 'G', 'ADD', '18486', '-0.05996', '-5.74', '9.625e-09', 'A']\n",
      "['13', 'rs2058727', '72390007', 'G', 'ADD', '18189', '-0.06082', '-5.765', '8.281e-09', 'A']\n"
     ]
    }
   ],
   "source": [
    "full = {}\n",
    "data = []\n",
    "for line in lines[1:]:\n",
    "    line = line.split()\n",
    "    if float(line[-2]) > 1e-8:\n",
    "        continue\n",
    "    else:\n",
    "        data.append([line[1], line[3], line[-2], line[-1]])\n",
    "        print(line)\n",
    "full[\"pheno98\"] = data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'pheno98': [['rs342909', 'T', '4.413e-11', 'C'],\n",
       "  ['rs11399396', 'T', '1.256e-11', 'TA'],\n",
       "  ['rs1674593', 'A', '9.269e-10', 'G'],\n",
       "  ['rs1674592', 'A', '9.461e-10', 'T'],\n",
       "  ['rs1702384', 'C', '9.461e-10', 'T'],\n",
       "  ['rs36008198', 'TGTA', '6.515e-10', 'T'],\n",
       "  ['rs1705839', 'T', '1.026e-09', 'G'],\n",
       "  ['rs1705840', 'T', '1.026e-09', 'G'],\n",
       "  ['rs1705841', 'T', '1.026e-09', 'G'],\n",
       "  ['rs34803973', 'A', '1.179e-09', 'C'],\n",
       "  ['rs386666824', 'AG', '9.927e-10', 'A'],\n",
       "  ['rs1702385', 'G', '1.026e-09', 'T'],\n",
       "  ['rs1702386', 'C', '1.208e-09', 'G'],\n",
       "  ['rs1705842', 'C', '1.161e-09', 'T'],\n",
       "  ['rs1705843', 'C', '1.208e-09', 'T'],\n",
       "  ['rs9810961', 'G', '1.15e-09', 'A'],\n",
       "  ['rs6799282', 'G', '1.123e-09', 'A'],\n",
       "  ['rs6799360', 'G', '1.202e-09', 'A'],\n",
       "  ['rs4541374', 'G', '1.279e-09', 'A'],\n",
       "  ['rs1702391', 'G', '1.603e-09', 'A'],\n",
       "  ['rs1674589', 'A', '1.823e-09', 'T'],\n",
       "  ['rs1674591', 'G', '1.823e-09', 'A'],\n",
       "  ['rs1627176', 'T', '1.603e-09', 'C'],\n",
       "  ['rs1705844', 'C', '1.288e-09', 'A'],\n",
       "  ['rs8179985', 'T', '1.271e-09', 'G'],\n",
       "  ['rs1380168', 'G', '1.357e-09', 'T'],\n",
       "  ['rs1380167', 'T', '1.376e-09', 'A'],\n",
       "  ['rs1702390', 'C', '1.327e-09', 'T'],\n",
       "  ['rs1702389', 'T', '1.327e-09', 'C'],\n",
       "  ['rs112461436', 'A', '9.226e-12', 'AAAGTT'],\n",
       "  ['rs1674587', 'T', '1.327e-09', 'A'],\n",
       "  ['rs1674586', 'G', '1.327e-09', 'A'],\n",
       "  ['rs1674585', 'T', '1.327e-09', 'A'],\n",
       "  ['rs1702387', 'G', '1.327e-09', 'A'],\n",
       "  ['rs1674584', 'T', '1.411e-09', 'C'],\n",
       "  ['rs1613380', 'A', '1.359e-09', 'C'],\n",
       "  ['rs9842222', 'T', '4.706e-09', 'C'],\n",
       "  ['rs9847565', 'G', '2.499e-10', 'C'],\n",
       "  ['rs12695824', 'A', '1.669e-12', 'T'],\n",
       "  ['rs57946797', 'A', '3.351e-11', 'T'],\n",
       "  ['rs4681329', 'A', '1.623e-12', 'G'],\n",
       "  ['rs6762490', 'A', '3.093e-12', 'T'],\n",
       "  ['rs4681330', 'C', '2.396e-12', 'T'],\n",
       "  ['rs6796791', 'T', '2.396e-12', 'C'],\n",
       "  ['rs9851764', 'C', '2.436e-12', 'A'],\n",
       "  ['rs12633109', 'T', '2.518e-12', 'G'],\n",
       "  ['rs303936', 'C', '7.183e-09', 'A'],\n",
       "  ['rs5804584', 'TA', '7.117e-09', 'T'],\n",
       "  ['rs5804585', 'GT', '8.578e-09', 'G'],\n",
       "  ['rs303933', 'G', '9.625e-09', 'A'],\n",
       "  ['rs2058727', 'G', '8.281e-09', 'A']]}"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/nas/longleaf/rhel8/apps/anaconda/2021.11/lib/python3.9/site-packages/requests/__init__.py:102: RequestsDependencyWarning: urllib3 (1.26.7) or chardet (5.2.0)/charset_normalizer (2.0.4) doesn't match a supported version!\n",
      "  warnings.warn(\"urllib3 ({}) or chardet ({})/charset_normalizer ({}) doesn't match a supported \"\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import openai\n",
    "import numpy as np\n",
    "openai.api_key = \"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\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "left inferior.parietal \n",
      " Anatomy: The left inferior parietal lobe is a section of the brain located above the occipital lobe and behind the frontal lobe. It is divided into two main sections, the supramarginal gyrus and the angular gyrus, both of which contribute to the brain's ability to process sensory information and coordinate spatial sense and navigation. \n",
      "\n",
      "Clinical function: The left inferior parietal lobule plays a pivotal role in various cognitive functions such as language processing, mathematical operations, body image, and spatial awareness. Damage to this area can lead to several disorders, such as Gerstmann's syndrome, which is characterized by writing, reading, and calculation impairments. Other possible conditions include left-right disorientation and finger agnosia.\n",
      "\n",
      "Genetic information: Genetic influences on the left inferior parietal like genes related to neurodevelopment and brain size can impact its function and structure. Specifically, genetic variants of the ASPM gene, which is crucial for brain development, have been linked to the structure of the inferior parietal lobule. Variations in genes affecting neurotransmitter levels, such as COMT, may also influence cognitive processes associated with this brain region. It's worth noting that current understanding in this field is limited and further genetic research is needed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "left inferior.temporal \n",
      " Anatomy: The left inferior temporal gyrus, part of the temporal lobe in the human brain, is one of three gyri in the temporal lobe. This gyrus is located underneath the middle temporal gyrus, and wraps around the lower end of lateral sulcus, which separates the frontal and parietal lobes from the temporal lobe. In close proximity to the auditory cortex, it plays a part in processing auditory input.\n",
      "\n",
      "Clinical function: The left inferior temporal gyrus plays a crucial role in the perception and recognition of visual stimuli, such as faces and objects. It is also thought to be involved in semantic processing and memory recall. Damage to this area may disrupt these processes, potentially leading to difficulties in recognizing and interpreting visual stimuli. A particular condition associated with dysfunction in this area is visual agnosia, the inability to recognize familiar objects or faces.\n",
      "\n",
      "Genetic Information: Genetic influence on the structure and function of the left inferior temporal gyrus is not yet completely understood. However, research has indicated possible linkage with certain heritable conditions such as schizophrenia and Alzheimer's disease, implicating specific genes (such as APOE4 for Alzheimer's) in the development and operation of this region. More research is needed to establish definite genetic links and impacts on its structure and function.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "left fusiform \n",
      " Anatomy: The left fusiform gyrus, part of the brain's cerebral cortex, is located in the fusiform area on the ventral side - or underside - of the brain, extending from the front temporal lobe to the back occipital lobe. It is critical for higher-level visual processing and plays a role in the recognition of faces, words, numbers, and colors. \n",
      "\n",
      "Clinical function: The left fusiform gyrus is traditionally associated with the processing of color information, word recognition, and face perception - a function which has earned it the nickname “fusiform face area”. Damage to the left fusiform can lead to prosopagnosia, the inability to recognize faces, or pure alexia, a condition characterized by an impaired ability to read. However, this area remains functionally intact for object recognition and spatial navigation.\n",
      "\n",
      "Genetic information: Several genes have been found to play a role in the structure and function of the fusiform gyrus, including the BDNF and COMT genes. Variations in these genes can influence brain structure and cognitive function, with certain polymorphisms being linked to exceptional memory performance, susceptibility to anxiety, or conversely, a higher risk for conditions such as Alzheimer's or schizophrenia. However, the exact genetic control over specific fusiform gyrus functions remains largely unknown.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "left cuneus \n",
      " Anatomy: The left cuneus is a part of the occipital lobe of the brain, located towards the back of the head. It is wedged in between the calcarine and parieto-occipital sulcus, situated on the medial surface of the cerebral hemisphere, just above the cerebellum. The left cuneus is one half of the larger, wing-shaped cuneus structure, with the right cuneus forming the other half.\n",
      "\n",
      "Clinical Function: The left cuneus primarily plays a role in basic visual processing. Given its positioning within the brain, it is primarily involved with processing visual information from the contralateral (right) visual field. It contributes to spatial awareness, the perception of size and shape, and aspects of motion detection. Lesions or damage to the cuneus area may result in visuospatial and perceptual deficits.\n",
      "\n",
      "Genetic Information: While no particular gene has been directly linked to the function of the left cuneus, research shows that genetic factors do play a role in the overall structure and function of the brain's occipital lobe, where the cuneus is housed. In general, the genetic makeup of an individual can influence brain morphology and can be associated with the susceptibility to neurological and psychiatric disorders.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "left entorhinal \n",
      " Anatomy: The left entorhinal cortex is located in the brain's medial temporal lobe. It borders the hippocampus, one of the key areas associated with memory and spatial functions. The entorhinal cortex is divided into two main areas: the lateral entorhinal area (LEA), associated with the olfactory cortex, and the medial entorhinal area (MEA), which is linked to the hippocampus. The left entorhinal cortex's exact structure includes several different layers of neurons.\n",
      "\n",
      "Clinical function: The entorhinal cortex is crucial for memory and navigation. It functions as the main interface between the hippocampus and the neocortex. The left side, in particular, tends to handle verbal memory, such as remembering words or events. Beyond memory, the left entorhinal cortex could also play roles in attention and perception, given its connections to other brain regions. Damage to this area can result in problems with memory and spatial recognition, presenting symptoms often found in Alzheimer's disease patients.\n",
      "\n",
      "Genetic Information: Several genetic variables may influence the left entorhinal cortex. APOE4, a genetic variation associated with an increased risk for Alzheimer's disease, can affect the entorhinal cortex's size. Furthermore, the FOX2 gene, associated with language skills, may also influence the left entorhinal cortex due to its role in verbal memory. However, much remains unknown about specific genetic influences on the entorhinal cortex, warranting further research.\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from openai_multi_client import OpenAIMultiClient\n",
    "\n",
    "# Remember to set the OPENAI_API_KEY environment variable to your API key\n",
    "api = OpenAIMultiClient(endpoint=\"chats\", data_template={\"model\": \"gpt-4\"})\n",
    "\n",
    "regions = [\"left cuneus\",\n",
    "\"left entorhinal\",\n",
    "\"left fusiform\",\n",
    "\"left inferior.parietal\",\n",
    "\"left inferior.temporal\"\n",
    "]\n",
    "\n",
    "def make_requests():\n",
    "    for region in regions:\n",
    "        api.request(data={\n",
    "            \"messages\": [{\"role\": \"user\", \"content\": f\"Can you tell me the Anatomy, Clinical function and genetic information of {region}? Please organize the answer with less than 200 words without any summary or conclusion. Only include the description items, organize the answer in three paragraphs, starting with Anatomy:, Clinical function:, genetic information:.\"}]\n",
    "        }, metadata={'region': region})\n",
    "\n",
    "\n",
    "api.run_request_function(make_requests)\n",
    "\n",
    "for result in api:\n",
    "    num = result.metadata['region']\n",
    "    response = result.response['choices'][0]['message']['content']\n",
    "    print(num, \"\\n\", response)\n",
    "    print(\"\\n\")\n",
    "    print(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "text = [{\"role\": \"user\", \"content\": \"Can you tell me the Anatomy, Clinical function and genetic information of left entorhinal? Please organize the answer with less than 200 words. Only include the description items, and separate each item with an asterisk (*)\"}]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'role': 'user',\n",
       "  'content': 'Can you tell me the Anatomy, Clinical function and genetic information of left entorhinal? Please organize the answer with less than 500 words'},\n",
       " {'role': 'user',\n",
       "  'content': 'Can you tell me the Anatomy, Clinical function and genetic information of left entorhinal? Please organize the answer with less than 500 words'},\n",
       " {'role': 'user',\n",
       "  'content': 'Can you tell me the Anatomy, Clinical function and genetic information of left entorhinal? Please organize the answer with less than 500 words'},\n",
       " {'role': 'user',\n",
       "  'content': 'Can you tell me the Anatomy, Clinical function and genetic information of left entorhinal? Please organize the answer with less than 500 words'},\n",
       " {'role': 'user',\n",
       "  'content': 'Can you tell me the Anatomy, Clinical function and genetic information of left entorhinal? Please organize the answer with less than 500 words'}]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "InvalidRequestError",
     "evalue": "[{'role': 'user', 'content': 'Can you tell me the Anatomy, Clinical function and genetic information of left entorhinal? Please organize the answer with less than 500 words'}] is not of type 'object' - 'messages.0'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mInvalidRequestError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[1;32m/nas/longleaf/home/runpeng/LLM/my/source_data/GPT/demo.ipynb 单元格 4\u001b[0m line \u001b[0;36m1\n\u001b[0;32m----> <a href='vscode-notebook-cell://ssh-remote%2Blongleaf/nas/longleaf/home/runpeng/LLM/my/source_data/GPT/demo.ipynb#W1sdnNjb2RlLXJlbW90ZQ%3D%3D?line=0'>1</a>\u001b[0m completion \u001b[39m=\u001b[39m openai\u001b[39m.\u001b[39;49mChatCompletion\u001b[39m.\u001b[39;49mcreate(\n\u001b[1;32m      <a href='vscode-notebook-cell://ssh-remote%2Blongleaf/nas/longleaf/home/runpeng/LLM/my/source_data/GPT/demo.ipynb#W1sdnNjb2RlLXJlbW90ZQ%3D%3D?line=1'>2</a>\u001b[0m   model\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mgpt-3.5-turbo\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m      <a href='vscode-notebook-cell://ssh-remote%2Blongleaf/nas/longleaf/home/runpeng/LLM/my/source_data/GPT/demo.ipynb#W1sdnNjb2RlLXJlbW90ZQ%3D%3D?line=2'>3</a>\u001b[0m   messages\u001b[39m=\u001b[39;49m text ,\n\u001b[1;32m      <a href='vscode-notebook-cell://ssh-remote%2Blongleaf/nas/longleaf/home/runpeng/LLM/my/source_data/GPT/demo.ipynb#W1sdnNjb2RlLXJlbW90ZQ%3D%3D?line=3'>4</a>\u001b[0m   max_tokens\u001b[39m=\u001b[39;49m \u001b[39m50\u001b[39;49m ,\n\u001b[1;32m      <a href='vscode-notebook-cell://ssh-remote%2Blongleaf/nas/longleaf/home/runpeng/LLM/my/source_data/GPT/demo.ipynb#W1sdnNjb2RlLXJlbW90ZQ%3D%3D?line=4'>5</a>\u001b[0m )\n\u001b[1;32m      <a href='vscode-notebook-cell://ssh-remote%2Blongleaf/nas/longleaf/home/runpeng/LLM/my/source_data/GPT/demo.ipynb#W1sdnNjb2RlLXJlbW90ZQ%3D%3D?line=6'>7</a>\u001b[0m \u001b[39mprint\u001b[39m(completion\u001b[39m.\u001b[39mchoices[\u001b[39m-\u001b[39m\u001b[39m1\u001b[39m]\u001b[39m.\u001b[39mmessage)\n",
      "File \u001b[0;32m~/.conda/envs/llama/lib/python3.11/site-packages/openai/api_resources/chat_completion.py:25\u001b[0m, in \u001b[0;36mChatCompletion.create\u001b[0;34m(cls, *args, **kwargs)\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[1;32m     24\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m---> 25\u001b[0m         \u001b[39mreturn\u001b[39;00m \u001b[39msuper\u001b[39;49m()\u001b[39m.\u001b[39;49mcreate(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m     26\u001b[0m     \u001b[39mexcept\u001b[39;00m TryAgain \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m     27\u001b[0m         \u001b[39mif\u001b[39;00m timeout \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m time\u001b[39m.\u001b[39mtime() \u001b[39m>\u001b[39m start \u001b[39m+\u001b[39m timeout:\n",
      "File \u001b[0;32m~/.conda/envs/llama/lib/python3.11/site-packages/openai/api_resources/abstract/engine_api_resource.py:155\u001b[0m, in \u001b[0;36mEngineAPIResource.create\u001b[0;34m(cls, api_key, api_base, api_type, request_id, api_version, organization, **params)\u001b[0m\n\u001b[1;32m    129\u001b[0m \u001b[39m@classmethod\u001b[39m\n\u001b[1;32m    130\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcreate\u001b[39m(\n\u001b[1;32m    131\u001b[0m     \u001b[39mcls\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    138\u001b[0m     \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mparams,\n\u001b[1;32m    139\u001b[0m ):\n\u001b[1;32m    140\u001b[0m     (\n\u001b[1;32m    141\u001b[0m         deployment_id,\n\u001b[1;32m    142\u001b[0m         engine,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    152\u001b[0m         api_key, api_base, api_type, api_version, organization, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mparams\n\u001b[1;32m    153\u001b[0m     )\n\u001b[0;32m--> 155\u001b[0m     response, _, api_key \u001b[39m=\u001b[39m requestor\u001b[39m.\u001b[39;49mrequest(\n\u001b[1;32m    156\u001b[0m         \u001b[39m\"\u001b[39;49m\u001b[39mpost\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m    157\u001b[0m         url,\n\u001b[1;32m    158\u001b[0m         params\u001b[39m=\u001b[39;49mparams,\n\u001b[1;32m    159\u001b[0m         headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    160\u001b[0m         stream\u001b[39m=\u001b[39;49mstream,\n\u001b[1;32m    161\u001b[0m         request_id\u001b[39m=\u001b[39;49mrequest_id,\n\u001b[1;32m    162\u001b[0m         request_timeout\u001b[39m=\u001b[39;49mrequest_timeout,\n\u001b[1;32m    163\u001b[0m     )\n\u001b[1;32m    165\u001b[0m     \u001b[39mif\u001b[39;00m stream:\n\u001b[1;32m    166\u001b[0m         \u001b[39m# must be an iterator\u001b[39;00m\n\u001b[1;32m    167\u001b[0m         \u001b[39massert\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(response, OpenAIResponse)\n",
      "File \u001b[0;32m~/.conda/envs/llama/lib/python3.11/site-packages/openai/api_requestor.py:299\u001b[0m, in \u001b[0;36mAPIRequestor.request\u001b[0;34m(self, method, url, params, headers, files, stream, request_id, request_timeout)\u001b[0m\n\u001b[1;32m    278\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mrequest\u001b[39m(\n\u001b[1;32m    279\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[1;32m    280\u001b[0m     method,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    287\u001b[0m     request_timeout: Optional[Union[\u001b[39mfloat\u001b[39m, Tuple[\u001b[39mfloat\u001b[39m, \u001b[39mfloat\u001b[39m]]] \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[1;32m    288\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Tuple[Union[OpenAIResponse, Iterator[OpenAIResponse]], \u001b[39mbool\u001b[39m, \u001b[39mstr\u001b[39m]:\n\u001b[1;32m    289\u001b[0m     result \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mrequest_raw(\n\u001b[1;32m    290\u001b[0m         method\u001b[39m.\u001b[39mlower(),\n\u001b[1;32m    291\u001b[0m         url,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    297\u001b[0m         request_timeout\u001b[39m=\u001b[39mrequest_timeout,\n\u001b[1;32m    298\u001b[0m     )\n\u001b[0;32m--> 299\u001b[0m     resp, got_stream \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_interpret_response(result, stream)\n\u001b[1;32m    300\u001b[0m     \u001b[39mreturn\u001b[39;00m resp, got_stream, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mapi_key\n",
      "File \u001b[0;32m~/.conda/envs/llama/lib/python3.11/site-packages/openai/api_requestor.py:710\u001b[0m, in \u001b[0;36mAPIRequestor._interpret_response\u001b[0;34m(self, result, stream)\u001b[0m\n\u001b[1;32m    702\u001b[0m     \u001b[39mreturn\u001b[39;00m (\n\u001b[1;32m    703\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_interpret_response_line(\n\u001b[1;32m    704\u001b[0m             line, result\u001b[39m.\u001b[39mstatus_code, result\u001b[39m.\u001b[39mheaders, stream\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m\n\u001b[1;32m    705\u001b[0m         )\n\u001b[1;32m    706\u001b[0m         \u001b[39mfor\u001b[39;00m line \u001b[39min\u001b[39;00m parse_stream(result\u001b[39m.\u001b[39miter_lines())\n\u001b[1;32m    707\u001b[0m     ), \u001b[39mTrue\u001b[39;00m\n\u001b[1;32m    708\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    709\u001b[0m     \u001b[39mreturn\u001b[39;00m (\n\u001b[0;32m--> 710\u001b[0m         \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_interpret_response_line(\n\u001b[1;32m    711\u001b[0m             result\u001b[39m.\u001b[39;49mcontent\u001b[39m.\u001b[39;49mdecode(\u001b[39m\"\u001b[39;49m\u001b[39mutf-8\u001b[39;49m\u001b[39m\"\u001b[39;49m),\n\u001b[1;32m    712\u001b[0m             result\u001b[39m.\u001b[39;49mstatus_code,\n\u001b[1;32m    713\u001b[0m             result\u001b[39m.\u001b[39;49mheaders,\n\u001b[1;32m    714\u001b[0m             stream\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    715\u001b[0m         ),\n\u001b[1;32m    716\u001b[0m         \u001b[39mFalse\u001b[39;00m,\n\u001b[1;32m    717\u001b[0m     )\n",
      "File \u001b[0;32m~/.conda/envs/llama/lib/python3.11/site-packages/openai/api_requestor.py:775\u001b[0m, in \u001b[0;36mAPIRequestor._interpret_response_line\u001b[0;34m(self, rbody, rcode, rheaders, stream)\u001b[0m\n\u001b[1;32m    773\u001b[0m stream_error \u001b[39m=\u001b[39m stream \u001b[39mand\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39merror\u001b[39m\u001b[39m\"\u001b[39m \u001b[39min\u001b[39;00m resp\u001b[39m.\u001b[39mdata\n\u001b[1;32m    774\u001b[0m \u001b[39mif\u001b[39;00m stream_error \u001b[39mor\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39m200\u001b[39m \u001b[39m<\u001b[39m\u001b[39m=\u001b[39m rcode \u001b[39m<\u001b[39m \u001b[39m300\u001b[39m:\n\u001b[0;32m--> 775\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandle_error_response(\n\u001b[1;32m    776\u001b[0m         rbody, rcode, resp\u001b[39m.\u001b[39mdata, rheaders, stream_error\u001b[39m=\u001b[39mstream_error\n\u001b[1;32m    777\u001b[0m     )\n\u001b[1;32m    778\u001b[0m \u001b[39mreturn\u001b[39;00m resp\n",
      "\u001b[0;31mInvalidRequestError\u001b[0m: [{'role': 'user', 'content': 'Can you tell me the Anatomy, Clinical function and genetic information of left entorhinal? Please organize the answer with less than 500 words'}] is not of type 'object' - 'messages.0'"
     ]
    }
   ],
   "source": [
    "completion = openai.ChatCompletion.create(\n",
    "  model=\"gpt-3.5-turbo\",\n",
    "  messages= text ,\n",
    "  max_tokens= 50 ,\n",
    ")\n",
    "\n",
    "print(completion.choices[-1].message)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<OpenAIObject chat.completion id=chatcmpl-8CYFzBnX5MMmBfNsbkFoM179amMK4 at 0x7fcc6092d370> JSON: {\n",
       "  \"id\": \"chatcmpl-8CYFzBnX5MMmBfNsbkFoM179amMK4\",\n",
       "  \"object\": \"chat.completion\",\n",
       "  \"created\": 1698002311,\n",
       "  \"model\": \"gpt-3.5-turbo-0613\",\n",
       "  \"choices\": [\n",
       "    {\n",
       "      \"index\": 0,\n",
       "      \"message\": {\n",
       "        \"role\": \"assistant\",\n",
       "        \"content\": \"The left entorhinal cortex is an important region of the brain that plays a critical role in memory and spatial navigation. It is located in the medial temporal lobe, specifically the rhinal sulcus, and is closely connected to other structures involved\"\n",
       "      },\n",
       "      \"finish_reason\": \"length\"\n",
       "    }\n",
       "  ],\n",
       "  \"usage\": {\n",
       "    \"prompt_tokens\": 67,\n",
       "    \"completion_tokens\": 50,\n",
       "    \"total_tokens\": 117\n",
       "  }\n",
       "}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "completion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "list = [[\"pheno1\", text], [\"pheno2\", text]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([['pheno1',\n",
       "        \"Anatomy: \\n\\nThe left entorhinal cortex (LEC), an important part of the limbic system, is situated in the medial temporal lobe of the brain. This brain area acts as a main interface between the hippocampus and neocortex. The LEC, with the right entorhinal cortex, is the primary interface between the hippocampus and neocortex. It is additionally responsible for key functions tied to memory, navigation, perception of time, and integration of sensory inputs. \\n\\nClinical Function:\\n\\nThe LEC has been implicated in a variety of functions, including episodic memory, spatial processing, and odor identification. Recent studies have indicated that it may play a crucial role in the encoding, consolidation, and retrieval of memories. The LEC is also associated with processing spatial data and the recognition of environmental cues vital for navigation.\\n\\nChanges in this cortex are seen in several neurodegenerative disorders. Early Alzheimer's disease reveals significant damage in the entorhinal cortex before spreading to adjacent regions. This may account for early symptoms such as difficulty remembering recent events or recognizing familiar locations. \\n\\nGenetic Information:\\n\\nThe understanding of the genetic makeup of LEC is not comprehensive yet. Several genetic variations could be linked to the LEC's broad range of functions and its vulnerability to neurodegenerative diseases. \\n\\nAlterations in genes such as APOE Ɛ4 (apolipoprotein E), presenilin 1 (PSEN1), and presenilin 2 (PSEN2), have been reported to be linked with Alzheimer's disease, which is known to impact areas including the entorhinal cortex. \\n\\nBrain-derived neurotrophic factor (BDNF), a gene that regulates neuronal survival and growth, has also been suggested to influence the LEC's function. In particular, the Met allele of the BDNF Val66Met polymorphism, which is related to reduced activity-dependent secretion of BDNF, has been associated with altered volume and activation in the entorhinal cortex. \\n\\nThe study of genes that influence the LEC will aid in understanding the mechanisms behind neuronal function and disease in this part of the brain. This might drive the development of therapeutic approaches targeted at the genetics influencing the LEC's function. However, further research is required to comprehend fully the genetic dynamics challenging the LEC and resultant disorders.\"],\n",
       "       ['pheno2',\n",
       "        \"Anatomy: \\n\\nThe left entorhinal cortex (LEC), an important part of the limbic system, is situated in the medial temporal lobe of the brain. This brain area acts as a main interface between the hippocampus and neocortex. The LEC, with the right entorhinal cortex, is the primary interface between the hippocampus and neocortex. It is additionally responsible for key functions tied to memory, navigation, perception of time, and integration of sensory inputs. \\n\\nClinical Function:\\n\\nThe LEC has been implicated in a variety of functions, including episodic memory, spatial processing, and odor identification. Recent studies have indicated that it may play a crucial role in the encoding, consolidation, and retrieval of memories. The LEC is also associated with processing spatial data and the recognition of environmental cues vital for navigation.\\n\\nChanges in this cortex are seen in several neurodegenerative disorders. Early Alzheimer's disease reveals significant damage in the entorhinal cortex before spreading to adjacent regions. This may account for early symptoms such as difficulty remembering recent events or recognizing familiar locations. \\n\\nGenetic Information:\\n\\nThe understanding of the genetic makeup of LEC is not comprehensive yet. Several genetic variations could be linked to the LEC's broad range of functions and its vulnerability to neurodegenerative diseases. \\n\\nAlterations in genes such as APOE Ɛ4 (apolipoprotein E), presenilin 1 (PSEN1), and presenilin 2 (PSEN2), have been reported to be linked with Alzheimer's disease, which is known to impact areas including the entorhinal cortex. \\n\\nBrain-derived neurotrophic factor (BDNF), a gene that regulates neuronal survival and growth, has also been suggested to influence the LEC's function. In particular, the Met allele of the BDNF Val66Met polymorphism, which is related to reduced activity-dependent secretion of BDNF, has been associated with altered volume and activation in the entorhinal cortex. \\n\\nThe study of genes that influence the LEC will aid in understanding the mechanisms behind neuronal function and disease in this part of the brain. This might drive the development of therapeutic approaches targeted at the genetics influencing the LEC's function. However, further research is required to comprehend fully the genetic dynamics challenging the LEC and resultant disorders.\"]],\n",
       "      dtype='<U2366')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.array(list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = ['pheno1', 'left caudal anterior cingulate', \"Anatomy: The left caudal anterior cingulate (cACC) is one of many sub-regions of the larger anterior cingulate cortex found in the brain. It's located in the medial (middle) portion of the brain, near the front. The ACC in general consists of several areas with distinct functions that are differentiated by cytoarchitecture and neurochemical profiles, among other characteristics. The cACC is identified as part of the limbic lobe, which is involved in emotion formation and processing, learning, and memory.\\n\\nClinical function: Clinically, the left cACC plays both broad and specific roles in emotion control, decision-making, and reward anticipation. It is particularly notable for its involvement in the experience and regulation of pain. Clinically, it's known to play a role in chronic pain conditions, and has also been associated with depression and anxiety, amongst other mental health disorders. Activation in this area of the brain is consistently found in tasks requiring conflict monitoring, response selection, and error detection, showing its intricate involvement in executive functions.\\n\\nGenetic information: The anatomy and function of the left cACC, like all brain structures, are influenced by our genetic makeup. However, connecting this region to specific genes is complicated. It is important to remark that, the molecular and cellular mechanisms in the cACC that contribute to neuropsychiatric disorders are beginning to emerge. Investigations surrounding the cACC often reference genes that are related to neurotransmitter function, such as glutamate and GABA, or those involved in synaptic plasticity. Variability in these genes can lead to differences in the structure or function of the cACC, which, in turn, can influence susceptibility to various neurological or psychiatric conditions.\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Anatomy: The left caudal anterior cingulate (cACC) is one of many sub-regions of the larger anterior cingulate cortex found in the brain. It's located in the medial (middle) portion of the brain, near the front. The ACC in general consists of several areas with distinct functions that are differentiated by cytoarchitecture and neurochemical profiles, among other characteristics. The cACC is identified as part of the limbic lobe, which is involved in emotion formation and processing, learning, and memory.**Clinical function: Clinically, the left cACC plays both broad and specific roles in emotion control, decision-making, and reward anticipation. It is particularly notable for its involvement in the experience and regulation of pain. Clinically, it's known to play a role in chronic pain conditions, and has also been associated with depression and anxiety, amongst other mental health disorders. Activation in this area of the brain is consistently found in tasks requiring conflict monitoring, response selection, and error detection, showing its intricate involvement in executive functions.**Genetic information: The anatomy and function of the left cACC, like all brain structures, are influenced by our genetic makeup. However, connecting this region to specific genes is complicated. It is important to remark that, the molecular and cellular mechanisms in the cACC that contribute to neuropsychiatric disorders are beginning to emerge. Investigations surrounding the cACC often reference genes that are related to neurotransmitter function, such as glutamate and GABA, or those involved in synaptic plasticity. Variability in these genes can lead to differences in the structure or function of the cACC, which, in turn, can influence susceptibility to various neurological or psychiatric conditions.\""
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "line = data[2].replace('\\n','*')\n",
    "line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savetxt('data.csv', data, delimiter=',', fmt='%s')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "could not convert string to float: 'pheno1'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_2925031/428764354.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloadtxt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'data.csv'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdelimiter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m','\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/nas/longleaf/rhel8/apps/anaconda/2021.11/lib/python3.9/site-packages/numpy/lib/npyio.py\u001b[0m in \u001b[0;36mloadtxt\u001b[0;34m(fname, dtype, comments, delimiter, converters, skiprows, usecols, unpack, ndmin, encoding, max_rows, like)\u001b[0m\n\u001b[1;32m   1144\u001b[0m         \u001b[0;31m# converting the data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1145\u001b[0m         \u001b[0mX\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1146\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mread_data\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_loadtxt_chunksize\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1147\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mX\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1148\u001b[0m                 \u001b[0mX\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nas/longleaf/rhel8/apps/anaconda/2021.11/lib/python3.9/site-packages/numpy/lib/npyio.py\u001b[0m in \u001b[0;36mread_data\u001b[0;34m(chunk_size)\u001b[0m\n\u001b[1;32m    995\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    996\u001b[0m             \u001b[0;31m# Convert each value according to its column and store\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 997\u001b[0;31m             \u001b[0mitems\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mconv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mval\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mconv\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconverters\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvals\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    998\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    999\u001b[0m             \u001b[0;31m# Then pack it according to the dtype's nesting\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nas/longleaf/rhel8/apps/anaconda/2021.11/lib/python3.9/site-packages/numpy/lib/npyio.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    995\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    996\u001b[0m             \u001b[0;31m# Convert each value according to its column and store\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 997\u001b[0;31m             \u001b[0mitems\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mconv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mval\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mconv\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconverters\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvals\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    998\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    999\u001b[0m             \u001b[0;31m# Then pack it according to the dtype's nesting\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nas/longleaf/rhel8/apps/anaconda/2021.11/lib/python3.9/site-packages/numpy/lib/npyio.py\u001b[0m in \u001b[0;36mfloatconv\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m    732\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;34m'0x'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    733\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mfloat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfromhex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 734\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mfloat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    735\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    736\u001b[0m     \u001b[0mtyp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: could not convert string to float: 'pheno1'"
     ]
    }
   ],
   "source": [
    "np.loadtxt('data.csv', delimiter=',')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "llama",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
